### ISSUE TITLE
Stopping export of metrics after some time of work

### ISSUE THREAD

**Issue by romaadereyko** [jvstein/bitcoin-prometheus-exporter#33]
Container is working, but prometheus don't see metrics after 1 week of working. Logs are stopping at this moment. this problem repeat on 7 different machines with different parameters

**Comment by jvstein** [comment]
@romaadereyko I need more information to help.

Are you able to view the metrics page in your browser or via `curl`? This will depend on how you are running the exporter, but typically something like `curl http://localhost:9332/metrics/` should return the values.

If that works, but Prometheus can't return the metrics, then it's likely that Prometheus isn't scraping the exporter. The exporter does all of it's work at request time, so you won't see anything in the logs until a request is made.

**Comment by romaadereyko** [comment]
Yep, curl view all metrics and they are change after every request. I’ll
try to screenshot this

Всего наилучшего
С уважением
Адерейко Р.М.
+375 29 3 800 787 А1
+375 33 3 800 787 МТС


Пн, 2 дек. 2024 г. в 22:08, Jeff Stein ***@***.***>:

**Comment by jvstein** [comment]
@romaadereyko Oh, so the metrics are collected successfully for a week and then they stop?

**Comment by romaadereyko** [comment]
Yes, it does.


Вт, 3 дек. 2024 г. в 21:30, Jeff Stein ***@***.***>:

**Comment by jvstein** [comment]
Does the exporter still respond with updated values via `curl` when they stop in Prometheus? Does the exporter return an error? How long does the request take?

If the exporter still responds, I would start looking at:
  - The logs of your prometheus installation, maybe with `--log.level=debug`.
  - The prometheus `up` and `scrape_duration_seconds` metrics from the prometheus server itself. A `up == 0` would indicate problems contacting the exporter and a high `scrape_duration_seconds` might indicate a timeout when collecting metrics.

**Comment by romaadereyko** [comment]
No, exporter respond with not updated metrics
I'll setup debug level & I'll see metrics when container wouldn't respond with updated metrics

**Comment by romaadereyko** [comment]
Exporter respond with not updated metrics. After restart exporter responds with updated metrics
[Exception.json](https://github.com/user-attachments/files/18056763/Exception.json)

**Comment by jvstein** [comment]
The `Connection reset by peer` means something is interrupting the connection during the request.

Do you have a reverse proxy in front of the exporter? Is there a configurable timeout on it? The timestamps look close, but I don't know about how that's all logged.

I would recommend trying a different version of Python to see if it's reproducible. Either way, this seems like a defect in the runtime environment or the `wsgiref` web server and not something I can address here.
